In [ ]:
#Libraries
from sklearn import datasets
import pandas as pd
import statsmodels.api as sm

#import gaussian Naive Bayes model
from sklearn.naive_bayes import GaussianNB
from sklearn import metrics
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score,recall_score,precision_score,f1_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report,roc_auc_score

import warnings
warnings.filterwarnings('ignore')
import numpy as np


In [ ]:
#IRIS DATASET
iriss = datasets.load_iris()
iris = pd.DataFrame(iriss.data)
iris.columns=iriss.feature_names
iris['Species']=iriss.target
iris.columns=iris.columns.str.replace(" ","")
iris.columns=iris.columns.str.replace("(","")
iris.columns=iris.columns.str.replace(")","")
iris.head()

sepallengthcm  sepalwidthcm  petallengthcm  petalwidthcm  Species
0            5.1           3.5            1.4           0.2        0
1            4.9           3.0            1.4           0.2        0
2            4.7           3.2            1.3           0.2        0
3            4.6           3.1            1.5           0.2        0
4            5.0           3.6            1.4           0.2        0

In [ ]:
#
iris['Species'].value_counts()



Species
0    50
1    50
2    50
Name: count, dtype: int64

from above output we see that is balanced(each species has equal no.of rows). hence, we use accuracy as metric

inference: The data is balanced

In [ ]:
#mean for every species
#Distribution (mean) of independent columns respect to dependent column

iris.groupby('Species').mean().round(2)

sepallengthcm  sepalwidthcm  petallengthcm  petalwidthcm
Species                                                          
0                 5.01          3.43           1.46          0.25
1                 5.94          2.77           4.26          1.33
2                 6.59          2.97           5.55          2.03

In [ ]:
iris.describe()

sepallengthcm  sepalwidthcm  petallengthcm  petalwidthcm     Species
count     150.000000    150.000000     150.000000    150.000000  150.000000
mean        5.843333      3.057333       3.758000      1.199333    1.000000
std         0.828066      0.435866       1.765298      0.762238    0.819232
min         4.300000      2.000000       1.000000      0.100000    0.000000
25%         5.100000      2.800000       1.600000      0.300000    0.000000
50%         5.800000      3.000000       4.350000      1.300000    1.000000
75%         6.400000      3.300000       5.100000      1.800000    2.000000
max         7.900000      4.400000       6.900000      2.500000    2.000000

In [ ]:
iris.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   sepallengthcm  150 non-null    float64
 1   sepalwidthcm   150 non-null    float64
 2   petallengthcm  150 non-null    float64
 3   petalwidthcm   150 non-null    float64
 4   Species        150 non-null    int64  
dtypes: float64(4), int64(1)
memory usage: 6.0 KB


Inference: All columns are numeric. There are no null values. no. of rows is 150.

In [ ]:
#Independent varaible
#Separate input and output
Independent_Variable_Base_Set=iris[iris.columns[0:4]]
Independent_Variable_Base_Set.head()

sepallengthcm  sepalwidthcm  petallengthcm  petalwidthcm
0            5.1           3.5            1.4           0.2
1            4.9           3.0            1.4           0.2
2            4.7           3.2            1.3           0.2
3            4.6           3.1            1.5           0.2
4            5.0           3.6            1.4           0.2

In [ ]:
#Dependent var
Dependent_Variable=iris[iris.columns[-1:iris.columns.size]]
Dependent_Variable.head()

Species
0        0
1        0
2        0
3        0
4        0

In [ ]:
#Split dataset

#Model 1
#Lets start with selecting one varaible
Independent_Variable_Set_v1=iris[iris.columns[0:1]]
X_train,X_test,y_train,y_test=train_test_split(Independent_Variable_Set_v1,Dependent_Variable,test_size=0.3,random_state=21)

In [ ]:
#Create gaussian Classifier
model= GaussianNB()

#train the model
result= model.fit(X_train,y_train)



In [ ]:
dir(result)

['__abstractmethods__',
 '__annotations__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setstate__',
 '__sizeof__',
 '__sklearn_clone__',
 '__sklearn_tags__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_abc_impl',
 '_build_request_for_signature',
 '_check_X',
 '_check_feature_names',
 '_check_n_features',
 '_doc_link_module',
 '_doc_link_template',
 '_doc_link_url_param_generator',
 '_estimator_type',
 '_get_default_requests',
 '_get_doc_link',
 '_get_metadata_request',
 '_get_param_names',
 '_get_tags',
 '_joint_log_likelihood',
 '_more_tags',
 '_parameter_constraints',
 '_partial_fit',
 '_repr_html_',
 '_repr_html_inner',
 '_repr_mimebundle_',
 '_update_mean_variance',
 '_validate_data',
 '_validat

In [ ]:
#Model Prediciton
print("Sample Prediction of model 1")
pred=result.predict(X_test)
model_prediction=pd.DataFrame(pred.round(2),columns=['Prediction'])
print(model_prediction['Prediction'].head())

Sample Prediction of model 1
0    1
1    0
2    0
3    0
4    1
Name: Prediction, dtype: int64


In [ ]:
#Test Set Target Column Distribution
print("\nTest Set Distribution")
print(y_test['Species'].value_counts())


Test Set Distribution
Species
1    16
2    15
0    14
Name: count, dtype: int64


In [ ]:
#predicted set target column distributuion
print("\nPredicted Set Distribution")
print(model_prediction['Prediction'].value_counts())


Predicted Set Distribution
Prediction
1    19
0    14
2    12
Name: count, dtype: int64


Inference: We cant infer anything with a single column. Instead, now we will use all the columns

In [ ]:
print("MODEL 1 PERFOMANCE")
y_pred=model_prediction[['Prediction']]
model_1_accuracy = accuracy_score(y_test,y_pred)
print("\nModel 1 Accuracy: ",model_1_accuracy)

model_1_precision=precision_score(y_test,y_pred,average='micro')
print("Model 1 Precision: ",model_1_precision)

model_1_recall=recall_score(y_test,y_pred,average='micro')
print("Model 1 Recall: ",model_1_recall)

model_1_f1_score=f1_score(y_test,y_pred,average='micro')
print("Model 1 F1 Score: ",model_1_f1_score)

print("\nConfusion Matrix,Model 1")
model_1_cm=confusion_matrix(y_test,y_pred)
print(model_1_cm)

print("\nClassification Report")
print(classification_report(y_test,y_pred))

print("Inferencee : Good Fit, can we make it better?")

MODEL 1 PERFOMANCE

Model 1 Accuracy:  0.6666666666666666
Model 1 Precision:  0.6666666666666666
Model 1 Recall:  0.6666666666666666
Model 1 F1 Score:  0.6666666666666666

Confusion Matrix,Model 1
[[13  1  0]
 [ 1 10  5]
 [ 0  8  7]]

Classification Report
              precision    recall  f1-score   support

           0       0.93      0.93      0.93        14
           1       0.53      0.62      0.57        16
           2       0.58      0.47      0.52        15

    accuracy                           0.67        45
   macro avg       0.68      0.67      0.67        45
weighted avg       0.67      0.67      0.66        45

Inferencee : Good Fit, can we make it better?


In [ ]:
#Model 2
#All features
X_train,X_test,y_train,y_test=train_test_split(Independent_Variable_Base_Set,Dependent_Variable,test_size=0.3,random_state=21)
#Train model using training sets
result=model.fit(X_train,y_train)

#model prediction
print("Sample Prediction of model 2")
pred=result.predict(X_test)
model_prediction=pd.DataFrame(pred.round(2),columns=['Prediction'])
print(model_prediction['Prediction'].head())

#test set target column distribution
print("\nTest Set Distribution")
print(y_test['Species'].value_counts())

#predicted set target column distribution
print("\nPredicted Set Distribution")
print(model_prediction['Prediction'].value_counts())
#

Sample Prediction of model 2
0    1
1    0
2    0
3    0
4    1
Name: Prediction, dtype: int64

Test Set Distribution
Species
1    16
2    15
0    14
Name: count, dtype: int64

Predicted Set Distribution
Prediction
2    17
1    14
0    14
Name: count, dtype: int64


In [ ]:
print("MODEL 2 PERFROMANCE")
y_pred=model_prediction[['Prediction']]
model_2_accuracy = accuracy_score(y_test,y_pred)
print("\nModel 2 Accuracy: ",model_2_accuracy)

model_2_precision=precision_score(y_test,y_pred,average='micro')
print("Model 2 Precision: ",model_2_precision)

model_2_recall=recall_score(y_test,y_pred,average='micro')
print("Model 2 Recall: ",model_2_recall)

model_2_f1_score=f1_score(y_test,y_pred,average='micro')
print("Model 2 F1 Score: ",model_2_f1_score)

print("\nConfusion Matrix,Model 1")
model_2_cm=confusion_matrix(y_test,y_pred)
print(model_2_cm)

print("\nClassification Report")
print(classification_report(y_test,y_pred))

print("Inference : Good Fit, can we make it better?")

MODEL 2 PERFOMANCE

Model 2 Accuracy:  0.9555555555555556
Model 2 Precision:  0.9555555555555556
Model 2 Recall:  0.9555555555555556
Model 2 F1 Score:  0.9555555555555556

Confusion Matrix,Model 1
[[14  0  0]
 [ 0 14  2]
 [ 0  0 15]]

Classification Report
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        14
           1       1.00      0.88      0.93        16
           2       0.88      1.00      0.94        15

    accuracy                           0.96        45
   macro avg       0.96      0.96      0.96        45
weighted avg       0.96      0.96      0.96        45

Inferencee : Good Fit, can we make it better?
